gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/mfa/MFAEnrollEndpoint.java

```diff
@@ -161,90 +169,89 @@ private void renderPage(RoutingContext routingContext) {
      * @param session current session
      * @return list of Factor object
      */
-    private List<Factor> factorsToRender(List<Factor> factors, Session session) {
+    private List<Factor> factorsToRender(List<Factor> factors, Session session, Client client, RoutingContext context) {
         // if an alternative factor ID has been set, only display this one
         final String alternativeFactorId = session.get(ConstantKeys.ALTERNATIVE_FACTOR_ID_KEY);
+        var mfaContext = new MfaFilterContext(context, client, factorManager);
+        var applicationFactors = ofNullable(client.getFactorSettings()).orElseGet(FactorSettings::new);
         if (alternativeFactorId != null && !alternativeFactorId.isEmpty()) {
             // check if this alternative factor still exists
             Optional<Factor> optionalFactor = factors.stream()
-                    .filter(factor -> alternativeFactorId.equals(factor.getId()))
+                    .filter(factor -> alternativeFactorId.equals(factor.getId()) && mfaContext.matchFactorRule(factor.getId(), applicationFactors, ruleEngine))
```

Comment:
**issue**: the `matchFactorRule` method provide factors which match a rule or have no rule but also the default factor. This is not good because the default factor need to be provided only if no factor is matching a rule.